### PR TITLE
Node testing mock overhaul

### PR DESCRIPTION
**Pull request by bmahall**
This PR makes the following changes to msal-node testing:--

TestFixtures.ts : Common mocks for usage , ex: Authority Factory, Telemetry Manager, etc.
 
For Client test classes:-
-Removes jest.mock("@azure/msal-common")
-Uses mock Authority & AuthorityFactory-createDiscoveredInstance's mock from 'test-fixtures.ts'
-Spies on AuthorizationCodeClient from common
-Mock implementation for various TokenClients 

For Cache test classes:-
-Added more tests based on the Coverage report.

 Storage.spec.ts:-
Tests supporting->
i) Change event being emitted when ChangeEmitter is registered
ii) Coverage for Catch blocks for setter and getter for Account, AccessToken, IdToken, RefreshToken

TokenCache.spec.ts:-
Tests supporting->
i) Edge cases like deserializing an empty cache, empty KVStore in case of empty Cache.

### CODE REVIEW

File: lib/msal-node/test/cache/Storage.spec.ts [hunk 1]
```diff
@@ -1,9 +1,11 @@
-import { LogLevel, Logger, AccountEntity, CacheManager, AccessTokenEntity, AuthorityMetadataEntity } from '@azure/msal-common';
+import { LogLevel, Logger, AccountEntity, CacheManager, AccessTokenEntity, AuthorityMetadataEntity, ThrottlingEntity, IdTokenEntity, RefreshTokenClient, RefreshTokenEntity } from '@azure/msal-common';
 import { JsonCache, InMemoryCache } from './../../src/cache/serializer/SerializerTypes';
 import { Deserializer } from './../../src/cache/serializer/Deserializer';
 import { NodeStorage } from '../../src/cache/NodeStorage';
 import { version, name } from '../../package.json';
 import { DEFAULT_CRYPTO_IMPLEMENTATION, DEFAULT_OPENID_CONFIG_RESPONSE, TEST_CONSTANTS } from '../utils/TestConstants';
+import { hasUncaughtExceptionCaptureCallback } from 'process';
```
Comment (bmahall) [inline lib/msal-node/test/cache/Storage.spec.ts]: - [x] Unused?

File: lib/msal-node/test/cache/Storage.spec.ts [hunk 2]
```diff
@@ -169,9 +187,69 @@ describe("Storage tests for msal-node: ", () => {
...
+        const invalidAccessToken = nodeStorage.getAccessTokenCredential(invalidAccessTokenKey);
+
         expect(fetchedAccessToken).toEqual(accessToken);
+        expect(invalidAccessToken).toBeNull();
     });
 
+    it('setIdTokenCredential() and getIdTokenCredential() tests', () => {
+        const nodeStorage = new NodeStorage(logger, clientId, DEFAULT_CRYPTO_IMPLEMENTATION);
+
+        //                  uid1.utid1-login.windows.net-idtoken-mock_client_id-samplerealm-
```
Comment (bmahall) [inline lib/msal-node/test/cache/Storage.spec.ts]: - [x] remove comment

File: lib/msal-node/test/cache/Storage.spec.ts [hunk 3]
```diff
@@ -169,9 +187,69 @@ describe("Storage tests for msal-node: ", () => {
...
+        const newMockEntityData = {
+            homeAccountId: 'uid1.utid1',
+            environment: 'login.windows.net',
+            credentialType: 'IdToken',
+            clientId: 'mock_client_id',
+            secret: 'an access token',
+            realm: 'samplerealm',
+        };
+
+        // <home_account_id*>-\<environment>-<credential_type>-<client_id>-<realm\*>-<target\*>-<scheme\*>
```
Comment (bmahall) [inline lib/msal-node/test/cache/Storage.spec.ts:208]: - [ ] remove comment

File: lib/msal-node/test/cache/Storage.spec.ts [hunk 4]
```diff
@@ -169,9 +187,69 @@ describe("Storage tests for msal-node: ", () => {
...
+        const idToken = CacheManager.toObject(new IdTokenEntity(), newMockEntityData);
+
+        nodeStorage.setIdTokenCredential(idToken);
+
+        require('fs').writeFileSync("temp.json", JSON.stringify(nodeStorage.getCache(), undefined, 2))
```
Comment (bmahall) [inline lib/msal-node/test/cache/Storage.spec.ts]: - [x] remove debug code!

File: lib/msal-node/test/cache/TokenCache.spec.ts [hunk 5]
```diff
@@ -57,8 +69,7 @@ describe("TokenCache tests", () => {
     });
 
     it("TokenCache.mergeRemovals removes entities from the cache, but does not remove other entities", async () => {
-        // TokenCache should not remove unrecognized entities from JSON file, even if they
-        // are deeply nested, and should write them back out
+        // TokenCache should not remove unrecognized entities from JSON file
```
Comment (bmahall) [inline lib/msal-node/test/cache/TokenCache.spec.ts]: - [x] put the existing comment back - unintentional change

File: lib/msal-node/test/cache/TokenCache.spec.ts [hunk 6]
```diff
@@ -95,17 +106,54 @@ describe("TokenCache tests", () => {
...
+        expect(msalCommon.TokenCacheContext).toHaveBeenCalled();
         expect(accounts.length).toBe(1);
         expect(require('./cache-test-files/temp-cache.json')).toEqual(require('./cache-test-files/cache-unrecognized-entities.json'));
 
         // try and clean up
         try {
             await fs.unlink(cachePath);
         } catch (err) {
-            if (err.code == "ENOENT") {
+            if ((err as any).code == "ENOENT") {
```
Comment (bmahall) [inline lib/msal-node/test/cache/TokenCache.spec.ts]: - [x] maybe not needed?

File: lib/msal-node/test/client/ConfidentialClientApplication.spec.ts [hunk 7]
```diff
@@ -1,34 +1,22 @@
...
-jest.mock('@azure/msal-common');
+import * as msalCommon from '@azure/msal-common';
+import { ClientCredentialRequest } from '../../src/request/ClientCredentialRequest';
+import { OnBehalfOfRequest } from '../../src';
+import { getMsalCommonAutoMock } from '../utils/MockUtils';
 
-mocked(StringUtils.isEmpty).mockImplementation((str) => {
-    return (typeof str === "undefined" || !str || 0 === str.length);
-});
+// jest.mock('@azure/msal-common');
```
Comment (bmahall) [inline lib/msal-node/test/client/ConfidentialClientApplication.spec.ts]: - [x] remove

File: lib/msal-node/test/client/ConfidentialClientApplication.spec.ts [hunk 8]
```diff
@@ -60,7 +48,15 @@ describe('ConfidentialClientApplication', () => {
...
         };
 
-        mocked(AuthorityFactory.createDiscoveredInstance).mockReturnValue(Promise.resolve(authority));
+        setupAuthorityFactory_createDiscoveredInstance_mock();
+        const mockAuthCodeClientInstance = {
+            includeRedirectUri: false,
+
+            acquireToken: jest.fn()
+        }
+        // const { AuthorizationCodeClient: MockAuthorizationCodeClient } = jest.genMockFromModule<typeof msalCommon>("@azure/msal-common")
```
Comment (bmahall) [inline lib/msal-node/test/client/ConfidentialClientApplication.spec.ts]: - [ ] remove

File: lib/msal-node/test/client/ConfidentialClientApplication.spec.ts [hunk 9]
```diff
@@ -70,13 +66,27 @@ describe('ConfidentialClientApplication', () => {
...
             refreshToken: TEST_CONSTANTS.REFRESH_TOKEN,
         };
 
-        mocked(AuthorityFactory.createDiscoveredInstance).mockReturnValue(Promise.resolve(authority));
+        setupAuthorityFactory_createDiscoveredInstance_mock();
+
+
+        const { RefreshTokenClient: mockRefreshTokenClient } = getMsalCommonAutoMock();
+        // const mockRefreshTokenClient = getMsalCommonAutoMock().RefreshTokenClient;
+        // const { RefreshTokenClient } = getMsalCommonAutoMock();
```
Comment (bmahall) [inline lib/msal-node/test/client/ConfidentialClientApplication.spec.ts]: - [ ] remove

File: lib/msal-node/test/client/ConfidentialClientApplication.spec.ts [hunk 10]
```diff
@@ -135,25 +161,4 @@ describe('ConfidentialClientApplication', () => {
         );
     });
 
-    test('acquireTokenByClientCredential handles AuthErrors as expected', async () => {
```
Comment (bmahall) [inline lib/msal-node/test/client/ConfidentialClientApplication.spec.ts]: - [x] add this back

File: lib/msal-node/test/client/PublicClientApplication.spec.ts [hunk 11]
```diff
@@ -1,42 +1,23 @@
...
-    AuthorityFactory,
-    AuthorizationCodeClient,
-    DeviceCodeClient,
-    RefreshTokenClient,
-    UsernamePasswordClient,
-    ClientConfiguration,
-    ProtocolMode,
-    Logger,
-    LogLevel
+    DeviceCodeClient, ClientConfiguration, AuthenticationResult,
```
Comment (bmahall) [inline lib/msal-node/test/client/PublicClientApplication.spec.ts]: - [ ] reformat/lint

File: lib/msal-node/test/client/PublicClientApplication.spec.ts [hunk 12]
```diff
@@ -74,40 +59,76 @@ describe('PublicClientApplication', () => {
...
-        mocked(AuthorityFactory.createDiscoveredInstance).mockReturnValue(Promise.resolve(authority));
+
+
+        // @ts-ignore
+        const mockDeviceCodeClient: DeviceCodeClient = {
+            acquireToken: jest.fn(),
+            createTokenRequestHeaders: jest.fn(),
+            executePostToTokenEndpoint: jest.fn(),
+            updateAuthority: jest.fn(),
+        }
```
Comment (bmahall) [inline lib/msal-node/test/client/PublicClientApplication.spec.ts]: - [x] remove/unused

File: lib/msal-node/test/client/PublicClientApplication.spec.ts [hunk 13]
```diff
@@ -163,6 +192,8 @@ describe('PublicClientApplication', () => {
             refreshToken: TEST_CONSTANTS.REFRESH_TOKEN,
         };
 
+        //Mocking Authority using AuthorityFactory's mock, but not RefreshToken Client ????
+
         const authorityMock = mocked(AuthorityFactory.createDiscoveredInstance);
         authorityMock.mockResolvedValue(authority);
```
Comment (bmahall) [inline lib/msal-node/test/client/PublicClientApplication.spec.ts]: - [ ] import from `test-fixtures.ts` instead

File: change/@azure-msal-node-95afe8ab-6566-4464-a5dc-a45498275294.json [hunk 14]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
```
Comment (jasonnutter) [inline change/@azure-msal-node-95afe8ab-6566-4464-a5dc-a45498275294.json]: Given this is just changing tests, I think this could be a `none` change (patch at most).

File: change/@azure-msal-node-95afe8ab-6566-4464-a5dc-a45498275294.json [hunk 15]
```diff
@@ -0,0 +1,7 @@
...
+  "comment": "Update node testing to overhaul msal-common mocking and add new tests",
```
Comment (jasonnutter) [inline change/@azure-msal-node-95afe8ab-6566-4464-a5dc-a45498275294.json]:
```suggestion
  "comment": "Update node testing to overhaul msal-common mocking and add new tests #4391",
```

File: lib/msal-node/test/cache/Storage.spec.ts [hunk 16]
```diff
@@ -50,6 +50,16 @@ describe("Storage tests for msal-node: ", () => {
         expect(Object.keys(inMemoryCache.accessTokens).length).toBe(0);
     });
 
+    it('emits a change event when changeEmitter is registered', () => {
```
Comment (sameerag) [inline lib/msal-node/test/cache/Storage.spec.ts:53]: +1 for adding this.I assume this also catches whenever this counter is increased when not needed.

File: lib/msal-node/test/cache/TokenCache.spec.ts [hunk 17]
```diff
@@ -57,7 +68,7 @@ describe("TokenCache tests", () => {
     });
 
     it("TokenCache.mergeRemovals removes entities from the cache, but does not remove other entities", async () => {
-        // TokenCache should not remove unrecognized entities from JSON file, even if they
+        // TokenCache should not remove unrecognized entities from JSON file, even if they are
```
Comment (sameerag) [inline lib/msal-node/test/cache/TokenCache.spec.ts]: nit: "are" is repeated twice

**Review by sameerag — APPROVED** [review]
Added two minor comments. Looks good. We should focus on coverage in the follow up PR.

**Review by samuelkubai — APPROVED** [review]
This looks good to me @bmahall, great work.

File: lib/msal-node/test/cache/Storage.spec.ts (hunk 16 above)
Comment (bmahall) [reply lib/msal-node/test/cache/Storage.spec.ts:53]: Yes, this would catch if the counter is increases more than expected number .
